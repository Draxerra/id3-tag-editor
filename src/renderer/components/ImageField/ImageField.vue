<template lang="pug">
    .ImageField(@dragover.prevent=""
                @drop.prevent="handleDrop")
        v-input.ImageFieldInput.mt-0
            .ImageFieldSlot
                .v-label.ImageFieldLabel {{ label }}
                .ImageFieldPlaceholder.ml-2.mt-2(v-if="!images.length") {{ placeholder }}
                img(v-else
                    v-for="image in images"
                    :src="image"
                    width="200").ml-2.mt-2.ImageFieldContainer
        v-btn(@click="showFileDialog" color="primary").ImageFieldUpdateBtn.mt-0.mb-3 Select Image
        v-btn(@click="removeImage" color="error").ImageFieldRemoveBtn.mt-0.mb-3 Remove Image
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Model, Prop } from "vue-property-decorator";

import VBtn from "vuetify/es5/components/VBtn";
import VInput from "vuetify/es5/components/VInput";
import VTextField from "vuetify/es5/components/VTextField";

@Component({
    components: {
        VBtn,
        VInput,
        VTextField,
    },
})
export default class AlbumArtField extends Vue {
    @Model("images-updated", { type: Array, required: true }) private images!: string[];
    @Prop({type: String}) private defaultImage!: string | undefined;
    @Prop({type: String, default: "Image"}) private label!: string;
    @Prop({type: String, default: "Drag Image Here"}) private placeholder!: string;

    private async handleDrop(event: DragEvent) {
        event.stopPropagation();
        const filteredFiles = Array.from(event.dataTransfer.files).filter((file) =>
            file.type.includes("image"),
        );
        if (filteredFiles.length) {
            try {
                const images = await this.getImages(filteredFiles);
                this.$emit("images-updated", images);
            } catch (e) {
                this.$emit("error", e);
            }
        } else {
            this.$emit("error", new Error("File type must be an image."));
        }
    }

    private getImages(files: File[]) {
        return Promise.all(files.map((file) => {
            return new Promise<string>((resolve, reject) => {
                const reader = new FileReader();
                reader.addEventListener("error", (e) => reject(e));
                reader.addEventListener("load", () => {
                    const data = new Uint8Array(reader.result as ArrayBuffer);
                    resolve(`data:${file.type};base64,${btoa(String.fromCharCode.apply(null, data))}`);
                });
                reader.readAsArrayBuffer(file);
            });
        }));
    }

    private showFileDialog() {
        const input = document.createElement("input");
        input.type = "file";
        input.accept = "image/*";
        input.click();
        input.addEventListener("change", async (event) => {
            event.preventDefault();
            const target = event.target as HTMLInputElement;
            if (target.files) {
                try {
                    const images = await this.getImages(Array.from(target.files));
                    this.$emit("images-updated", images);
                } catch (e) {
                    this.$emit("error", e);
                }
            }
        });
    }

    private removeImage() {
        this.$emit("images-updated", this.defaultImage ? [this.defaultImage] : []);
    }
}
</script>

<style lang="stylus" scoped>
.ImageFieldContainer {
    border: 1px solid black;
}
.ImageFieldPlaceholder {
    @extend .ImageFieldContainer;
    height: 200px;
    width: 200px;
    display: flex;
    align-items: center;
    justify-content: center;
}
</style>
