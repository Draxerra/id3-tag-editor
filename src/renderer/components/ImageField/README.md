# Image Field

## Functionality

The image field component is an input component, which allows a user to: add, update and remove images. Images may be added via file dialog or drag/drop.

### Props

`<image-field>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `default-image` | String | The default base64 encoded image value.
| `images` | string[] | The array of base64 encoded images. Can be bound to using `v-model`.
| `label` | string *optional* | The form label.
| `placeholder` | string *optional* | The form placeholder.

### Events

`<image-field>` emits the following events:

| event | args type | description
| --- | --- | ---
| `error` | Error | An error object.
| `images-updated` | string[] | The updated array of base64 encoded images. Emitted automatically when using `v-model`.

### Example

```html
<image-field :images="images"
             label="Album Art"
             placeholder="Drop Image Here" />
```

```ts
import Vue from 'vue';
import { Component } from "vue-property-decorator";

import ImageField from '~/src/renderer/components/ImageField/ImageField.vue';

@Component({
    components: {
        ImageField,
    },
})
export default class GenericComponent extends Vue {
    private images = ["data:image/;base64,"];
}
```