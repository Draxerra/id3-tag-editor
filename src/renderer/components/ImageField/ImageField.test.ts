import { config, mount } from "@vue/test-utils";
import ImageField from "./ImageField.vue";

import Vue from "vue";
import Vuetify from "vuetify/es5/components/Vuetify";

Vue.use(Vuetify);
config.logModifiedComponents = false;

const blankImage = "data:image/png;base64,AAAAAAAAAAA=";

describe("Image Field Component", () => {

    it("is a Vue instance", () => {
        const wrapper = mount(ImageField, { propsData: { images: [blankImage] } });
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it("snapshot has not changed", () => {
        const wrapper = mount(ImageField, { propsData: { images: [blankImage] } });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("placeholder snapshot has not changed", () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("displays placeholder with empty images array", () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        expect(wrapper.find(".ImageFieldContainer").exists()).toBeFalsy();
        expect(wrapper.find(".ImageFieldPlaceholder").exists()).toBeTruthy();
    });

    it("displays placeholder prop", () => {
        const placeholder = "<keep>";
        const wrapper = mount(ImageField, { propsData: { images: [], placeholder } });
        expect(wrapper.find(".ImageFieldPlaceholder").text()).toEqual(placeholder);
    });

    it("drop event with valid file emits images-updated with array of base64 images", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test", { type: "image/png" })];

        const eventListener = jest.fn();
        spyOn(window, "FileReader" as any).and.returnValue({
            addEventListener: eventListener,
            readAsArrayBuffer: jest.fn(),
            result: buf,
        });
        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });

        const loadCall = eventListener.mock.calls.find((call) => call[0] === "load");
        if (loadCall) {
            loadCall[1]();
            await wrapper.vm.$nextTick();
            expect(wrapper.emitted("error")).toBeUndefined();
            expect(wrapper.emitted("images-updated")[0]).toEqual([["data:image/png;base64,AAAAAAAAAAA="]]);
        }
    });

    it("drop event with file reader error emits error", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        const files = [new File([new ArrayBuffer(8)], "test", { type: "image/png" })];
        const err = new Error("Invalid File");

        const eventListener = jest.fn();
        spyOn(window, "FileReader" as any).and.returnValue({
            addEventListener: eventListener,
            readAsArrayBuffer: jest.fn(),
        });
        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        const errorCall = eventListener.mock.calls.find((call) => call[0] === "error");
        if (errorCall) {
            errorCall[1](err);
            await wrapper.vm.$nextTick();
            expect(wrapper.emitted("images-updated")).toBeUndefined();
            expect(wrapper.emitted("error")[0]).toEqual([err]);
        }
    });

    it("drop event with other file type emits invalid file error", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test", { type: "text/csv" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });

        expect(wrapper.emitted("images-updated")).toBeUndefined();
        expect(wrapper.emitted("error")).toEqual([[new Error("File type must be an image.")]]);
    });

    it("clicking update image opens file dialog and emits images-updated with array of base64 images", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test", { type: "image/png" })];

        const inputEventListener = jest.fn();
        spyOn(document, "createElement").and.returnValue({
            addEventListener: inputEventListener,
            click: jest.fn(),
        });

        const fileReaderEventListener = jest.fn();
        spyOn(window, "FileReader" as any).and.returnValue({
            addEventListener: fileReaderEventListener,
            readAsArrayBuffer: jest.fn(),
            result: buf,
        });

        wrapper.find(".ImageFieldUpdateBtn").trigger("click");
        const changeCall = inputEventListener.mock.calls.find((call) => call[0] === "change");
        if (changeCall) {
            changeCall[1]({
                preventDefault: jest.fn(),
                target: {
                    files,
                },
            });
            const loadCall = fileReaderEventListener.mock.calls.find((call) => call[0] === "load");
            if (loadCall) {
                loadCall[1]();
                await wrapper.vm.$nextTick();
                expect(wrapper.emitted("error")).toBeUndefined();
                expect(wrapper.emitted("images-updated")).toEqual([[["data:image/png;base64,AAAAAAAAAAA="]]]);
            }
        }
    });

    it("clicking update image opens file dialog and emits error with file reader error", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test", { type: "image/png" })];

        const inputEventListener = jest.fn();
        spyOn(document, "createElement").and.returnValue({
            addEventListener: inputEventListener,
            click: jest.fn(),
        });

        const fileReaderEventListener = jest.fn();
        spyOn(window, "FileReader" as any).and.returnValue({
            addEventListener: fileReaderEventListener,
            readAsArrayBuffer: jest.fn(),
        });

        wrapper.find(".ImageFieldUpdateBtn").trigger("click");
        const changeCall = inputEventListener.mock.calls.find((call) => call[0] === "change");
        if (changeCall) {
            changeCall[1]({
                preventDefault: jest.fn(),
                target: {
                    files,
                },
            });
            const errorCall = fileReaderEventListener.mock.calls.find((call) => call[0] === "error");
            if (errorCall) {
                const err = new Error("Invalid File");
                errorCall[1](err);
                await wrapper.vm.$nextTick();
                expect(wrapper.emitted("images-updated")).toBeUndefined();
                expect(wrapper.emitted("error")).toEqual([[err]]);
            }
        }
    });

    it("clicking update image opens file dialog and emits nothing with no files", async () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });

        const inputEventListener = jest.fn();
        spyOn(document, "createElement").and.returnValue({
            addEventListener: inputEventListener,
            click: jest.fn(),
        });

        wrapper.find(".ImageFieldUpdateBtn").trigger("click");
        const changeCall = inputEventListener.mock.calls.find((call) => call[0] === "change");
        if (changeCall) {
            await changeCall[1]({
                preventDefault: jest.fn(),
                target: {},
            });
            expect(wrapper.emitted("images-updated")).toBeUndefined();
            expect(wrapper.emitted("error")).toBeUndefined();
        }
    });

    it("emits images-updated on remove button click", () => {
        const wrapper = mount(ImageField, { propsData: { images: [] } });
        wrapper.find(".ImageFieldRemoveBtn").trigger("click");
        expect(wrapper.emitted("images-updated")).toEqual([[[]]]);
    });

    it("emits images-updated on remove button click with default image prop", () => {
        const wrapper = mount(ImageField, { propsData: { defaultImage: blankImage, images: [] } });
        wrapper.find(".ImageFieldRemoveBtn").trigger("click");
        expect(wrapper.emitted("images-updated")).toEqual([[[blankImage]]]);
    });

});
