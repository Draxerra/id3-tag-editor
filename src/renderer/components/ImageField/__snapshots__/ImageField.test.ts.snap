// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Field Component placeholder snapshot has not changed 1`] = `
<div class="ImageField">
  <div class="v-input ImageFieldInput mt-0">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="ImageFieldSlot">
          <div class="v-label ImageFieldLabel">Image</div>
          <div class="ImageFieldPlaceholder ml-2 mt-2">Drag Image Here</div>
        </div>
      </div>
      <div class="v-messages">
        <div class="v-messages__wrapper"></div>
      </div>
    </div>
  </div>
  <button type="button" class="ImageFieldUpdateBtn mt-0 mb-3 v-btn primary">
    <div class="v-btn__content">Update Image</div>
  </button>
  <button type="button" class="ImageFieldRemoveBtn mt-0 mb-3 v-btn error">
    <div class="v-btn__content">Remove Image</div>
  </button>
</div>
`;

exports[`Image Field Component snapshot has not changed 1`] = `
<div class="ImageField">
  <div class="v-input ImageFieldInput mt-0">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="ImageFieldSlot">
          <div class="v-label ImageFieldLabel">Image</div>
          <img src="data:image/png;base64,AAAAAAAAAAA=" width="200" class="ml-2 mt-2 ImageFieldContainer">
        </div>
      </div>
      <div class="v-messages">
        <div class="v-messages__wrapper"></div>
      </div>
    </div>
  </div>
  <button type="button" class="ImageFieldUpdateBtn mt-0 mb-3 v-btn primary">
    <div class="v-btn__content">Update Image</div>
  </button>
  <button type="button" class="ImageFieldRemoveBtn mt-0 mb-3 v-btn error">
    <div class="v-btn__content">Remove Image</div>
  </button>
</div>
`;
